File: shared/engine/rng.js
// ═══════════════════════════════════════════════════════════════════
//  Nova Edge — موتور مشترک: RNG با seed
//  همهٔ سه ران‌تایم (JS ورکر / Java پیپر / PHP پاکت‌ماین) از همین
//  الگوریتم استفاده می‌کنند تا رفتار بات‌ها قابل بازتولید و تست باشد.
//  الگوریتم: mulberry32 + xoshiro برای تنوع؛ خروجی یکسان در هر زبان.
// ═══════════════════════════════════════════════════════════════════

/** mulberry32 — کوچک، سریع، یکسان در Java/PHP/JS */
export function mulberry32(seed) {
  let a = seed >>> 0;
  return function () {
    a = (a + 0x6d2b79f5) >>> 0;
    let t = a;
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

export class Rng {
  constructor(seed = 1) {
    this.seed = (Number(seed) || 1) >>> 0;
    this._next = mulberry32(this.seed);
  }

  /** عدد اعشاری [0,1) */
  float() {
    return this._next();
  }

  /** عدد صحیح [min,max] شامل هر دو */
  int(min, max) {
    const lo = Math.ceil(Number(min) || 0);
    const hi = Math.floor(Number(max) || 0);
    if (hi <= lo) return lo;
    return lo + Math.floor(this._next() * (hi - lo + 1));
  }

  /** شانس وقوع با احتمال p */
  chance(p) {
    return this._next() < Number(p || 0);
  }

  /** انتخاب تصادفی از آرایه */
  pick(arr) {
    if (!arr || !arr.length) return null;
    return arr[Math.floor(this._next() * arr.length) % arr.length];
  }

  /** وزن‌دار: items=[{w:number,...}] */
  weighted(items, weightKey = 'w') {
    if (!items || !items.length) return null;
    let total = 0;
    for (const it of items) total += Math.max(0, Number(it[weightKey]) || 0);
    if (total <= 0) return items[0];
    let r = this._next() * total;
    for (const it of items) {
      r -= Math.max(0, Number(it[weightKey]) || 0);
      if (r <= 0) return it;
    }
    return items[items.length - 1];
  }

  /** گاوسی تقریبی (میانگین ۰، انحراف ۱) — برای خطای انسانی بات */
  gauss() {
    let u = 0;
    let v = 0;
    while (u === 0) u = this._next();
    while (v === 0) v = this._next();
    return Math.sqrt(-2 * Math.log(u)) * Math.cos(2 * Math.PI * v);
  }

  /** نرمال با میانگین و انحراف دلخواه */
  normal(mean, sd) {
    return Number(mean || 0) + this.gauss() * Number(sd || 1);
  }

  shuffle(arr) {
    const a = Array.isArray(arr) ? arr.slice() : [];
    for (let i = a.length - 1; i > 0; i--) {
      const j = Math.floor(this._next() * (i + 1));
      const t = a[i];
      a[i] = a[j];
      a[j] = t;
    }
    return a;
  }
}

/** هش رشته → عدد ۳۲ بیتی (برای seed بات از روی نام مچ) */
export function hashSeed(str) {
  let h = 2166136261 >>> 0;
  const s = String(str || '');
  for (let i = 0; i < s.length; i++) {
    h ^= s.charCodeAt(i);
    h = Math.imul(h, 16777619) >>> 0;
  }
  return h >>> 0 || 1;
}

export const clamp = (n, min, max) => Math.max(Number(min), Math.min(Number(max), Number(n) || 0));
export const round = (n, digits = 0) => {
  const f = Math.pow(10, digits);
  return Math.round((Number(n) || 0) * f) / f;
};

File: server/java/src/main/java/ir/novaedge/core/Rng.java
// ═══════════════════════════════════════════════════════════════════
//  Nova Edge — core/Rng.java (پورت دقیق shared/engine/rng.js)
//  جاوا int سی‌ودو بیتی signed است — دقیقاً همان سمانتیک |0 در JS.
// ═══════════════════════════════════════════════════════════════════
package ir.novaedge.core;

public final class Rng {

    private Rng() {
    }

    public static int u32(long x) {
        return (int) (x & 0xFFFFFFFFL);
    }

    /** معادل Math.round جاوااسکریپت (نصف به +∞) برای double */
    public static long jsRound(double x) {
        return (long) Math.floor(x + 0.5);
    }

    public static double roundTo(double n, int digits) {
        double f = Math.pow(10, digits);
        return jsRound(n * f) / f;
    }

    public static double clamp(double n, double min, double max) {
        double v = Double.isFinite(n) ? n : 0;
        return Math.max(min, Math.min(max, v));
    }

    /** mulberry32 — دنبالهٔ یکسان با JS/PHP */
    public static java.util.function.DoubleSupplier mulberry32(int seed) {
        final int[] a = {seed};
        return () -> {
            a[0] = a[0] + 0x6D2B79F5;
            int t = a[0];
            int a1 = t ^ (t >>> 15);
            int b1 = t | 1;
            t = a1 * b1; // ضرب int جاوا = Math.imul (wrap mod 2^32)
            int a2 = t ^ (t >>> 7);
            int b2 = t | 61;
            t ^= t + a2 * b2;
            return Integer.toUnsignedLong(t ^ (t >>> 14)) / 4294967296.0;
        };
    }

    /** هش FNV-1a مثل hashSeed در rng.js */
    public static int hashSeed(String s) {
        int h = 2166136261;
        for (int i = 0; i < s.length(); i++) {
            h ^= s.charAt(i);
            h *= 16777619;
        }
        return h == 0 ? 1 : h;
    }

    /** کلاس Rng با متدهای float/int/chance/pick/weighted/gauss/normal/shuffle */
    public static class Generator {
        private final java.util.function.DoubleSupplier next;

        public Generator(int seed) {
            this.next = mulberry32(seed == 0 ? 1 : seed);
        }

        public double float_() {
            return next.getAsDouble();
        }

        public int int_(double min, double max) {
            int lo = (int) Math.ceil(min);
            int hi = (int) Math.floor(max);
            if (hi <= lo) {
                return lo;
            }
            return lo + (int) Math.floor(next.getAsDouble() * (hi - lo + 1));
        }

        public boolean chance(double p) {
            return next.getAsDouble() < p;
        }

        public <T> T pick(java.util.List<T> arr) {
            if (arr == null || arr.isEmpty()) {
                return null;
            }
            return arr.get((int) Math.floor(next.getAsDouble() * arr.size()) % arr.size());
        }

        public double gauss() {
            double u = 0;
            double v = 0;
            while (u == 0) {
                u = next.getAsDouble();
            }
            while (v == 0) {
                v = next.getAsDouble();
            }
            return Math.sqrt(-2 * Math.log(u)) * Math.cos(2 * Math.PI * v);
        }

        public double normal(double mean, double sd) {
            return mean + gauss() * sd;
        }

        public <T> java.util.List<T> shuffle(java.util.List<T> arr) {
            java.util.List<T> a = new java.util.ArrayList<>(arr);
            for (int i = a.size() - 1; i > 0; i--) {
                int j = (int) Math.floor(next.getAsDouble() * (i + 1));
                T t = a.get(i);
                a.set(i, a.get(j));
                a.set(j, t);
            }
            return a;
        }
    }
}
